lib/Dialect/RTL/Ops.cpp

```diff
@@ -657,12 +657,53 @@ void AddOp::getCanonicalizationPatterns(OwningRewritePatternList &results,
         return success();
       }
 
+      // add(..., x, x) -> add(..., mul(x, 2))
+      if (inputs[size - 1] == inputs[size - 2]) {
+        SmallVector<Value, 2> mulInputs;
+
+        auto two = rewriter.create<ConstantOp>(
+            op.getLoc(), 2, op.getType().cast<IntegerType>());
+        mulInputs.push_back(inputs.back());
+        mulInputs.push_back(two);
+
+        auto mulOp = rewriter.create<rtl::MulOp>(op.getLoc(), mulInputs);
+
+        SmallVector<Value, 4> newOperands(inputs.drop_back(/*n=*/2));
+        newOperands.push_back(mulOp);
+        rewriter.replaceOpWithNewOp<AddOp>(op, op.getType(), newOperands);
+        return success();
+      }
+
+      auto mulOp = inputs[size - 1].getDefiningOp<rtl::MulOp>();
+      APInt multiplier;
+
+      // add(..., x, mul(x, c)) -> add(..., mul(x, c + 1))
+      if (mulOp && mulOp.inputs().size() == 2 &&
+          mulOp.inputs()[0] == inputs[size - 2] &&
+          matchPattern(mulOp.inputs()[1], m_RConstant(multiplier))) {
+
+        SmallVector<Value, 2> mulInputs(mulOp.inputs().drop_back());
+
+        auto numBits =
+            mulOp.getOperand(1).getType().cast<IntegerType>().getWidth();
+        APInt one(numBits, 1, /*isSigned=*/false);
```

Comment:
instead of getting the width from the type, I'd just get it from `multiplier.getWidth()` since you have it handy and they are guaranteed to be the same.